### PR TITLE
Upgrade TypeScript and improve invariant/warn helpers

### PR DESCRIPTION
**Pull request by kitten**
- Upgrade to TypeScript 3.7
- Make use of `asserts` on `invariant` and refactor
- Rename exchanges (`cacheExchange.ts` and `populateExchange.ts`)
- Update `invariant` and `warn` in `populateExchange.ts` to have entries in `help.md`

### CODE REVIEW

File: docs/help.md [hunk 1]
```diff
@@ -244,3 +244,28 @@ in the cache, which is then treated as a fragment match.
 
 When you pass an introspected schema to the cache, this warning will never be displayed
 as the cache can then do deterministic fragment matching using schema information.
+
+## (17) Invalid type <a id="17"></a>
```
Comment (andyrichardson) [inline docs/help.md]: Looks good :+1:

My two cents - I think having populate errors documented alongside graphcache might be confusing (separate exchanges, unrelated errors)

Comment (kitten) [reply docs/help.md]: I think it's fine-ish. I expect that some may become shared as we start sharing some utilities (if that ever becomes a thing) and I expect them to be looked up (via links in the error/log output) individually, so the user will only look at the linked section that they're currently interested in

File: src/helpers/help.ts [hunk 2]
```diff
@@ -50,7 +54,7 @@ export const invariant = (clause: any, message: string, code: ErrorCode) => {
   }
 };
 
-export const warn = (message: string, code: ErrorCode) => {
+export function warn(message: string, code: ErrorCode) {
```
Comment (andyrichardson) [inline src/helpers/help.ts]: preference for block functions?

Comment (kitten) [reply src/helpers/help.ts]: Everything with `asserts` must use the function keyword. Don’t ask me why, it doesn’t make sense 😂

So I converted `warn` as well just to make sure it’s consistent in this file

**Pull request merged**